File: challenges/01_fizzbuzz/cpp/fizzbuzz.cpp
// FizzBuzz - C++ 実装

#include <iostream>
#include <string>
#include <vector>
#include <ranges>
#include <algorithm>
#include <functional>
#include <cassert>

using namespace std;

// === 基本実装 ===
void fizzbuzz_basic(int n) {
    cout << "--- Basic Implementation ---" << endl;

    for (int i = 1; i <= n; ++i) {
        if (i % 15 == 0) {
            cout << "FizzBuzz" << endl;
        } else if (i % 3 == 0) {
            cout << "Fizz" << endl;
        } else if (i % 5 == 0) {
            cout << "Buzz" << endl;
        } else {
            cout << i << endl;
        }
    }
    cout << endl;
}

// === 関数型スタイル ===
string fizzbuzz_transform(int i) {
    if (i % 15 == 0) return "FizzBuzz";
    if (i % 3 == 0) return "Fizz";
    if (i % 5 == 0) return "Buzz";
    return to_string(i);
}

vector<string> fizzbuzz_functional(int n) {
    vector<string> result;
    result.reserve(n);

    for (int i = 1; i <= n; ++i) {
        result.push_back(fizzbuzz_transform(i));
    }

    return result;
}

void functional_demo() {
    cout << "--- Functional Style ---" << endl;

    auto result = fizzbuzz_functional(15);

    for (const auto& s : result) {
        cout << s << " ";
    }
    cout << endl << endl;
}

// === ranges (C++20) ===
void fizzbuzz_ranges() {
    cout << "--- Ranges (C++20) ---" << endl;

    auto result = views::iota(1, 16)
        | views::transform([](int i) -> string {
            if (i % 15 == 0) return "FizzBuzz";
            if (i % 3 == 0) return "Fizz";
            if (i % 5 == 0) return "Buzz";
            return to_string(i);
        });

    for (const auto& s : result) {
        cout << s << " ";
    }
    cout << endl << endl;
}

// === 文字列連結スタイル ===
void fizzbuzz_concat() {
    cout << "--- String Concatenation Style ---" << endl;

    for (int i = 1; i <= 15; ++i) {
        string s;
        if (i % 3 == 0) s += "Fizz";
        if (i % 5 == 0) s += "Buzz";
        if (s.empty()) s = to_string(i);
        cout << s << " ";
    }
    cout << endl << endl;
}

// === テンプレートメタプログラミング ===
template<int N>
struct FizzBuzz {
    static string get() {
        if constexpr (N % 15 == 0) return "FizzBuzz";
        else if constexpr (N % 3 == 0) return "Fizz";
        else if constexpr (N % 5 == 0) return "Buzz";
        else return to_string(N);
    }
};

void fizzbuzz_template() {
    cout << "--- Template Metaprogramming ---" << endl;

    cout << "1: " << FizzBuzz<1>::get() << endl;
    cout << "3: " << FizzBuzz<3>::get() << endl;
    cout << "5: " << FizzBuzz<5>::get() << endl;
    cout << "15: " << FizzBuzz<15>::get() << endl;

    cout << endl;
}

// === constexpr ===
constexpr auto fizzbuzz_constexpr(int i) {
    struct Result {
        bool is_fizzbuzz;
        bool is_fizz;
        bool is_buzz;
        int number;
    };

    return Result{
        i % 15 == 0,
        i % 3 == 0 && i % 5 != 0,
        i % 5 == 0 && i % 3 != 0,
        i
    };
}

void constexpr_demo() {
    cout << "--- constexpr ---" << endl;

    constexpr auto r3 = fizzbuzz_constexpr(3);
    constexpr auto r5 = fizzbuzz_constexpr(5);
    constexpr auto r15 = fizzbuzz_constexpr(15);

    static_assert(r3.is_fizz);
    static_assert(r5.is_buzz);
    static_assert(r15.is_fizzbuzz);

    cout << "Compile-time checks passed!" << endl;
    cout << endl;
}

// === ジェネレータ風 ===
class FizzBuzzGenerator {
public:
    class Iterator {
    public:
        using iterator_category = input_iterator_tag;
        using value_type = string;
        using difference_type = ptrdiff_t;
        using pointer = const string*;
        using reference = const string&;

        Iterator(int current, int end)
            : current_(current), end_(end) {
            if (current_ <= end_) update_value();
        }

        const string& operator*() const { return value_; }

        Iterator& operator++() {
            ++current_;
            if (current_ <= end_) update_value();
            return *this;
        }

        bool operator!=(const Iterator& other) const {
            return current_ != other.current_;
        }

    private:
        void update_value() {
            value_ = fizzbuzz_transform(current_);
        }

        int current_;
        int end_;
        string value_;
    };

    FizzBuzzGenerator(int start, int end) : start_(start), end_(end) {}

    Iterator begin() const { return Iterator(start_, end_); }
    Iterator end() const { return Iterator(end_ + 1, end_); }

private:
    int start_;
    int end_;
};

void generator_demo() {
    cout << "--- Generator Style ---" << endl;

    FizzBuzzGenerator gen(1, 15);

    for (const auto& s : gen) {
        cout << s << " ";
    }
    cout << endl << endl;
}

// === ルールテーブル ===
void fizzbuzz_rules() {
    cout << "--- Rules Table ---" << endl;

    vector<pair<int, string>> rules = {
        {15, "FizzBuzz"},
        {3, "Fizz"},
        {5, "Buzz"}
    };

    for (int i = 1; i <= 15; ++i) {
        string result;
        for (const auto& [divisor, word] : rules) {
            if (i % divisor == 0) {
                result = word;
                break;
            }
        }
        if (result.empty()) result = to_string(i);
        cout << result << " ";
    }
    cout << endl << endl;
}

// === テスト ===
void run_tests() {
    cout << "--- Tests ---" << endl;

    auto result = fizzbuzz_functional(15);

    // 基本ケース
    assert(result[0] == "1");
    assert(result[1] == "2");

    // Fizz (3の倍数)
    assert(result[2] == "Fizz");
    assert(result[5] == "Fizz");

    // Buzz (5の倍数)
    assert(result[4] == "Buzz");
    assert(result[9] == "Buzz");

    // FizzBuzz (15の倍数)
    assert(result[14] == "FizzBuzz");

    cout << "All tests passed!" << endl;
}

int main() {
    cout << "=== FizzBuzz Demo ===" << endl << endl;

    fizzbuzz_basic(15);
    functional_demo();
    fizzbuzz_ranges();
    fizzbuzz_concat();
    fizzbuzz_template();
    constexpr_demo();
    generator_demo();
    fizzbuzz_rules();
    run_tests();

    return 0;
}

File: challenges/01_fizzbuzz/typescript/fizzbuzz.ts
// FizzBuzz - TypeScript 実装
//
// 様々なアプローチで FizzBuzz を実装

console.log("=== FizzBuzz Demo ===\n");

// --- 基本実装 ---
function basicFizzBuzz(): void {
  console.log("--- Basic Implementation ---");

  function fizzBuzz(n: number): string {
    if (n % 15 === 0) return "FizzBuzz";
    if (n % 3 === 0) return "Fizz";
    if (n % 5 === 0) return "Buzz";
    return String(n);
  }

  for (let i = 1; i <= 15; i++) {
    console.log(fizzBuzz(i));
  }
  console.log();
}

// --- 配列とmap ---
function arrayFizzBuzz(): void {
  console.log("--- Array Map Implementation ---");

  const result = Array.from({ length: 15 }, (_, i) => i + 1).map((n) => {
    const fizz = n % 3 === 0 ? "Fizz" : "";
    const buzz = n % 5 === 0 ? "Buzz" : "";
    return fizz + buzz || String(n);
  });

  console.log(result.join(", "));
  console.log();
}

// --- ジェネレータ ---
function* fizzBuzzGenerator(limit: number): Generator<string> {
  for (let i = 1; i <= limit; i++) {
    const fizz = i % 3 === 0 ? "Fizz" : "";
    const buzz = i % 5 === 0 ? "Buzz" : "";
    yield fizz + buzz || String(i);
  }
}

function generatorFizzBuzz(): void {
  console.log("--- Generator Implementation ---");

  const gen = fizzBuzzGenerator(15);
  console.log([...gen].join(", "));
  console.log();
}

// --- 関数型 ---
function functionalFizzBuzz(): void {
  console.log("--- Functional Implementation ---");

  type Rule = [divisor: number, word: string];

  const rules: Rule[] = [
    [3, "Fizz"],
    [5, "Buzz"],
  ];

  const fizzBuzz = (n: number): string => {
    const result = rules
      .filter(([div]) => n % div === 0)
      .map(([, word]) => word)
      .join("");
    return result || String(n);
  };

  const output = Array.from({ length: 15 }, (_, i) => fizzBuzz(i + 1));
  console.log(output.join(", "));
  console.log();
}

// --- パターンマッチング風 ---
function patternMatchFizzBuzz(): void {
  console.log("--- Pattern Match Style ---");

  const fizzBuzz = (n: number): string => {
    const match: [boolean, boolean] = [n % 3 === 0, n % 5 === 0];

    switch (true) {
      case match[0] && match[1]:
        return "FizzBuzz";
      case match[0]:
        return "Fizz";
      case match[1]:
        return "Buzz";
      default:
        return String(n);
    }
  };

  for (let i = 1; i <= 15; i++) {
    process.stdout.write(fizzBuzz(i) + " ");
  }
  console.log("\n");
}

// --- 型安全な拡張版 ---
function typeSafeFizzBuzz(): void {
  console.log("--- Type-Safe Extended Version ---");

  interface FizzBuzzRule {
    divisor: number;
    word: string;
  }

  class FizzBuzzEngine {
    private rules: FizzBuzzRule[];

    constructor(rules: FizzBuzzRule[] = []) {
      this.rules = rules.sort((a, b) => a.divisor - b.divisor);
    }

    addRule(divisor: number, word: string): this {
      this.rules.push({ divisor, word });
      this.rules.sort((a, b) => a.divisor - b.divisor);
      return this;
    }

    generate(n: number): string {
      const result = this.rules
        .filter((rule) => n % rule.divisor === 0)
        .map((rule) => rule.word)
        .join("");
      return result || String(n);
    }

    *range(start: number, end: number): Generator<string> {
      for (let i = start; i <= end; i++) {
        yield this.generate(i);
      }
    }
  }

  const engine = new FizzBuzzEngine()
    .addRule(3, "Fizz")
    .addRule(5, "Buzz")
    .addRule(7, "Bazz");

  console.log("With Fizz(3), Buzz(5), Bazz(7):");
  const results = [...engine.range(1, 21)];
  console.log(results.join(", "));
  console.log();
}

// --- 非同期版 ---
async function asyncFizzBuzz(): Promise<void> {
  console.log("--- Async Implementation ---");

  async function* asyncFizzBuzzGen(limit: number): AsyncGenerator<string> {
    for (let i = 1; i <= limit; i++) {
      await new Promise((r) => setTimeout(r, 10));
      const fizz = i % 3 === 0 ? "Fizz" : "";
      const buzz = i % 5 === 0 ? "Buzz" : "";
      yield fizz + buzz || String(i);
    }
  }

  const results: string[] = [];
  for await (const value of asyncFizzBuzzGen(10)) {
    results.push(value);
  }
  console.log(results.join(", "));
  console.log();
}

// --- テスト ---
function runTests(): void {
  console.log("--- Tests ---");

  const fizzBuzz = (n: number): string => {
    if (n % 15 === 0) return "FizzBuzz";
    if (n % 3 === 0) return "Fizz";
    if (n % 5 === 0) return "Buzz";
    return String(n);
  };

  const tests: [number, string][] = [
    [1, "1"],
    [2, "2"],
    [3, "Fizz"],
    [4, "4"],
    [5, "Buzz"],
    [6, "Fizz"],
    [10, "Buzz"],
    [15, "FizzBuzz"],
    [30, "FizzBuzz"],
  ];

  let passed = 0;
  for (const [input, expected] of tests) {
    const result = fizzBuzz(input);
    if (result === expected) {
      passed++;
    } else {
      console.log(`FAIL: fizzBuzz(${input}) = "${result}", expected "${expected}"`);
    }
  }

  console.log(`${passed}/${tests.length} tests passed`);
}

// --- 実行 ---
async function main(): Promise<void> {
  basicFizzBuzz();
  arrayFizzBuzz();
  generatorFizzBuzz();
  functionalFizzBuzz();
  patternMatchFizzBuzz();
  typeSafeFizzBuzz();
  await asyncFizzBuzz();
  runTests();
}

main();
